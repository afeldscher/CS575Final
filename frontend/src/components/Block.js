import BlockDataContent from "./BlockDataContent";
import React from 'react';

class Block extends React.Component {
    constructor(props) {
        super(props);
        this.triggerMine = this.triggerMine.bind(this);
    }

    triggerMine() {
        this.props.mineFunction(this.props.id);
    }

    render() {
        let backgroundClass = "block col s11 m5 l5";
        let statusString = "";
        if (this.props.mined) {
            backgroundClass += " mined-block";
            statusString = "Good";
        } else {
            backgroundClass += " bad-block";
            statusString = "Bad"
        }
        return (
            <div
                className={backgroundClass}>
                <div>
                    <div className="block-header">
                        <span className="block-title">Block #{this.props.id + 1}</span><br/>
                        <span>Status: {statusString}</span>
                    </div>
                    <TextBoxInput id={"guid"} value={this.props.guid} label={"GUID"}/>
                    <DisabledTextAreaInput id={"parent"} value={this.props.parent} label={"Parent"}/>
                    <BlockDataContent id={this.props.id} checkId={this.props.checkId} checkItems={this.props.checkItems}
                                      checkCost={this.props.checkCost} checkTip={this.props.checkTip} onChangeFct={this.props.textareaOnChange}/>
                    <TextBoxInput id={"nonce"} value={this.props.nonce} label={"Nonce"}/>
                    <DisabledTextAreaInput id={"hash"} label={"Hash"} value={this.props.hash}/>
                    <button className="waves-effect waves-light btn mine-btn green darken-1"
                            onClick={this.triggerMine}>Mine
                    </button>
                </div>
            </div>)
    }
}

function TextBoxInput(props) {
    return (
        <div className="input-field">
            <input className="block-field" type="text" id={props.id} disabled={true} value={props.value}/>
            <label className="active">{props.label}</label>
        </div>);
}

function DisabledTextAreaInput(props) {
    return (
        <div className="input-field">
            <textarea readOnly className="materialize-textarea hash-textarea" id={props.id} value={props.value}/>
            <label className="active">{props.label}</label>
        </div>);
}

export default Block;