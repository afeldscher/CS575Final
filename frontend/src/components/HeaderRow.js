import React from 'react';

export class HeaderRow extends React.Component {
    DEFAULT_ZEROES = 4;
    DEFAULT_NUM_TRIES = 10000000;

    render() {
        return (
            <div>
                <div className=" header-row row">
                    <div className="col s4">
                        <a title="Click to add a block to the blockchain"
                           href="#addItemModal" className="green darken-1 waves-effect waves-light btn modal-trigger">
                            Add a block
                        </a>
                    </div>
                    <div className="col s4">
                        <div title="Specifies the difficulty for the hash solving algorithm" className="input-field">
                            <input onChange={this.props.resetBlockFunction} defaultValue={this.DEFAULT_ZEROES} min="1"
                                   max="32" id="numZeroes" type="number"/>
                            <label htmlFor="numZeroes">Number of Leading Zeroes</label>
                        </div>
                    </div>
                    <div className="col s4">
                        <div title="The maximum number of nonce values to try" className="input-field">
                            <input defaultValue={this.DEFAULT_NUM_TRIES} min="100" id="numTries" type="number"/>
                            <label htmlFor="numTries">Max Tries Per Mine</label>
                        </div>
                    </div>
                </div>
                <div className="modal" id="addItemModal">
                    <div className="modal-content">
                        <h5>Enter data for new block</h5>
                        <div className="input-field add-block-textarea-wrapper">
                            <AddBlockForm/>
                        </div>
                    </div>
                    <div className="modal-footer">
                        <a id="addItemModalButton" href="#" onClick={this.props.addBlockFunction}
                           className="modal-close waves-effect btn-flat">
                            Add
                        </a>
                    </div>
                </div>
                <div className="modal" id="infoModal">
                    <div className="modal-content">
                        <h4>About this Application</h4>
                        <p>This application is a blockchain simulator that uses a proof-of-work consensus algorithm.
                            The specific hashing function is based on the
                            <a href="https://en.bitcoin.it/wiki/Block_hashing_algorithm"> bitcoin hashing algorithm.</a>
                        </p>
                    </div>
                    <div className="modal-footer">
                        <a href="#" className="modal-close waves-effect btn-flat">Close</a>
                    </div>
                </div>
            </div>
        );
    }
}

function AddBlockForm() {
    return (
        <div>
            <div className="input-field">
                <input placeholder="Enter check ID" id="checkId" type="text"/>
                <label htmlFor="checkId">Check Id</label>
            </div>
            <div className="input-field">
                <textarea className="materialize-textarea" placeholder="Enter items on bill" id="checkItems"/>
                <label htmlFor="checkItems">Items</label>
            </div>
            <div className="input-field">
                <input placeholder="Enter total bill ($)" id="checkCost" step="any" type="number"/>
                <label htmlFor="checkCost">Total Bill</label>
            </div>
            <div className="input-field">
                <input placeholder="Enter tip" id="checkTip" type="number" step="any"/>
                <label htmlFor="checkTip">Tip</label>
            </div>
        </div>

    )
}

export default HeaderRow;