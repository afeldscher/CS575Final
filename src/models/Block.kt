package com.blockchain.models

import com.blockchain.hexStringToByteArray
import com.blockchain.isHex
import org.apache.commons.codec.digest.DigestUtils
import java.nio.BufferOverflowException
import java.nio.ByteBuffer


class Block(in_version: Int, in_parent_hash: String, in_data: String, in_sec_since_epoc: Int, in_target_zeros: Int, in_nonce: Int) {
    private val version: Int = in_version
    private val parent_hash: String = in_parent_hash
    private val data: String = in_data
    private val sec_since_epoc: Int = in_sec_since_epoc
    val target_zeros: Int = in_target_zeros
    var nonce: Int = in_nonce

    constructor(): this(0, "", "", 0, 0, 0) {}

    fun is_valid(): Boolean {
        return version >= 0 &&
                parent_hash.length == 64 &&
                parent_hash.isHex() &&
                sec_since_epoc >= 0 &&
                target_zeros >= 0
    }

    fun get_hash(): ByteArray {
        val data_hash = DigestUtils.sha256(data) //TODO extra nonce

        // https://en.bitcoin.it/wiki/Block_hashing_algorithm
        try {
            // --- Build up byte array to hash ---
            val hashdata = ByteBuffer.allocate(80)
            hashdata.putInt(version)
            hashdata.put(parent_hash.toUpperCase().hexStringToByteArray())
            hashdata.put(data_hash)
            hashdata.putInt(sec_since_epoc)
            hashdata.putInt(target_zeros)
            hashdata.putInt(nonce)

            // --- Build up byte array to hash ---
            //https://www.javacodemonk.com/md5-and-sha-256-in-java-kotlin-and-android-96ed9628s
            return DigestUtils.sha256(hashdata.array())

        } catch (e: BufferOverflowException) {
            println("Buffer Overflow! This is a bug!")
            e.printStackTrace()
            return ByteArray(0)
        }
    }
}


