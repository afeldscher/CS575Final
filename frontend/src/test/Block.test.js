import React from 'react';
import Block from '../components/Block';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import HeaderRow from "../components/HeaderRow";



describe('Test <Block>', () => {
    const idx = 1;
    const item = {
        parent: '0000',
        nonce: 100,
        hash: '1234567890abcdef',
        checkId: 'check-id',
        checkItems: 'check-items',
        checkCost: 10.05,
        checkTip: 1.05,
        mined: true
    };
    const mineBlockMock = jest.fn();
    const textareaOnChangeMock = jest.fn();

    it('renders without crashing', () => {
        shallow( <Block key={idx} id={idx} guid={item.guid} parent={item.parent} nonce={item.nonce}
                        hash={item.hash} mineFunction={mineBlockMock} mined={item.mined}
                        textareaOnChange={textareaOnChangeMock} checkId={item.checkId} checkItems={item.checkItems}
                        checkCost={item.checkCost} checkTip={item.checkTip}
        />);
    });

    it('matches snapshot', () => {
        const tree = renderer.create(
            <Block key={idx} id={idx} guid={item.guid} parent={item.parent} nonce={item.nonce}
                   hash={item.hash} mineFunction={mineBlockMock} mined={item.mined}
                   textareaOnChange={textareaOnChangeMock} checkId={item.checkId} checkItems={item.checkItems}
                   checkCost={item.checkCost} checkTip={item.checkTip}
            />,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});