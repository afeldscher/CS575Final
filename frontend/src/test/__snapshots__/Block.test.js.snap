// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Block> matches snapshot 1`] = `
<div
  className="block col s11 m5 l5 mined-block"
>
  <div>
    <div
      className="block-header"
    >
      <span
        className="block-title"
      >
        Block #
        2
      </span>
      <br />
      <span>
        Status: 
        Good
      </span>
    </div>
    <div
      className="input-field"
    >
      <input
        className="block-field"
        disabled={true}
        id="guid"
        type="text"
      />
      <label
        className="active"
      >
        GUID
      </label>
    </div>
    <div
      className="input-field"
    >
      <textarea
        className="materialize-textarea hash-textarea"
        id="parent"
        readOnly={true}
        value="0000"
      />
      <label
        className="active"
      >
        Parent
      </label>
    </div>
    <div>
      <div
        className="input-field"
      >
        <input
          className="block-field"
          id="checkId-1"
          onChange={[Function]}
          type="text"
          value="check-id"
        />
        <label
          className="active"
          htmlFor="checkId-1"
        >
          Check ID
        </label>
      </div>
      <div
        className="input-field"
      >
        <textarea
          className="materialize-textarea data-textarea"
          id="checkItems-1"
          onChange={[Function]}
        >
          check-items
        </textarea>
        <label
          className="active"
          htmlFor="checkItems-1"
        >
          Items
        </label>
      </div>
      <div
        className="input-field"
      >
        <input
          className="block-field"
          id="checkCost-1"
          onChange={[Function]}
          step="any"
          type="number"
          value={10.05}
        />
        <label
          className="active"
          htmlFor="checkCost-1"
        >
          Cost
        </label>
      </div>
      <div
        className="input-field"
      >
        <input
          className="block-field"
          id="checkTip-1"
          onChange={[Function]}
          step="any"
          type="number"
          value={1.05}
        />
        <label
          className="active"
          htmlFor="checkTip-1"
        >
          Tip
        </label>
      </div>
    </div>
    <div
      className="input-field"
    >
      <input
        className="block-field"
        disabled={true}
        id="nonce"
        type="text"
        value={100}
      />
      <label
        className="active"
      >
        Nonce
      </label>
    </div>
    <div
      className="input-field"
    >
      <textarea
        className="materialize-textarea hash-textarea"
        id="hash"
        readOnly={true}
        value="1234567890abcdef"
      />
      <label
        className="active"
      >
        Hash
      </label>
    </div>
    <button
      className="waves-effect waves-light btn mine-btn green darken-1"
      onClick={[Function]}
    >
      Mine
    </button>
  </div>
</div>
`;
